### ISSUE TITLE
Tensor algebra without constructing the full tensor

### ISSUE THREAD

**Issue by act65** [tensorly/tensorly#78]
I think this might fall under a feature request, unless you can already do it (please tell me how!?!).

It would be nice if we could do operations on the decomposed tensors, in their decomposed form.
(or does this sort of thing get done in the various compiler optimisations???)
 
```python
A = tf.random_normal([1,n])
B = tf.random_normal([1,n])

# where the tensor object contains the cores and factors, not the full tensor
tensor = tucker(tf.random_normal([n,n,n]), rank=[2, 2, 2])
reduce_tensor = multi_mode_dot(tensor, [A, B], modes=[0, 1])
```

In this case it is possible to evaluate the `multi_mode_dot` op without constructing the full tucker decomposed tensor. If the factors = [U, V, W] then we can simply evaluate something like W^T.(core x A.U x B.V).

**Comment by animakumar** [comment]
I agree. A general framework for implicitly dealing with tensors would be powerful. We should cover the broad class of representations (e.g CP, Tucker, TT, sparse, hierarchical etc). After a set of tensor operations, the framework should figure out the best representation for output tensor

**Comment by JeanKossaifi** [comment]
Absolutely, this is in the work :)

Currently we do not have a specific Tensor class, which makes the abstraction hard. However, I am working on functions to efficiently manipulate the decomposed tensors. 
Feel free to contribute any you may have!

**Comment by JeanKossaifi** [comment]
Partially addressed (for Kruskal/CP tensors) by #87

**Comment by JeanKossaifi** [comment]
We now have several functions for manipulating decomposed tensors for CP and Tucker, and will keep adding more.
